crates/e2e-testing/src/testcase.rs

```diff
@@ -0,0 +1,137 @@
+use crate::controller::Controller;
+use crate::metadata_extractor::AppMetadata;
+use crate::spin;
+use crate::utils;
+use anyhow::{Context, Result};
+use std::fs;
+use tokio::task;
+
+pub struct SkipCondition {
+    pub env: String,
+    pub reason: String,
+}
+
+impl SkipCondition {
+    pub fn skip(&self, controller: &dyn Controller) -> bool {
+        controller.name() == self.env
+    }
+}
+
+/// Represents a testcase
+pub struct TestCase {
+    /// name of the testcase
+    pub name: String,
+
+    /// name of the app under test
+    pub appname: Option<String>,
+
+    /// optional
+    /// template to use to create new app
+    pub template: Option<String>,
+
+    /// optional
+    /// template install args. appended to `spin install templates <template_install_args>
+    /// defaults to `--git https://github.com/fermyon/spin`
+    pub template_install_args: Option<Vec<String>>,
+
+    /// optional
+    /// plugins required for the testcase. e.g. js2wasm for js/ts tests
+    pub plugins: Option<Vec<String>>,
+
+    /// optional
+    /// conditions where this testcase should be skipped
+    /// currently only supports skipping based on controller name
+    pub skip_conditions: Option<Vec<SkipCondition>>,
+
+    /// optional
+    /// if provided, appended to `spin deploy` command
+    pub deploy_args: Option<Vec<String>>,
+
+    /// optional
+    /// if provided, executed as command line before running `spin build`
+    /// useful if some external script or process needs to run before `spin build`
+    /// e.g. `npm install` before running `spin build` for `js/ts` tests
+    pub pre_build_hooks: Option<Vec<Vec<String>>>,
+
+    /// assertions to run once the app is running
+    pub assertions: fn(app: &AppMetadata) -> Result<()>,
+}
+
+impl TestCase {
+    pub async fn run(&self, controller: &dyn Controller) -> Result<()> {
+        controller.name();
+
+        // evaluate the skip conditions specified in testcase config.
+        if let Some(skip_conditions) = &self.skip_conditions {
+            for skip_condition in skip_conditions {
+                if skip_condition.skip(controller) {
+                    return Ok(());
+                }
+            }
+        }
+
+        // install spin plugins if requested in testcase config
+        if let Some(plugins) = &self.plugins {
+            controller
+                .install_plugins(plugins.iter().map(|s| s as &str).collect())
+                .context("installing plugins")?;
+        }
+
+        let appname = match &self.appname {
+            Some(appname) => appname.to_owned(),
+            None => format!("{}-generated", self.template.as_ref().unwrap()),
+        };
+
+        let appdir = spin::appdir(appname.as_str());
+
+        // cleanup existing dir for testcase project code. cleaned up only if testcase is a template based test
+        if self.template.is_some() {
+            // appname = format!("{}-generated", &self.template.unwrap()).as_mut();
```

Comment:
did you mean to leave these comments in here?

Reply:
no :)

Reply:
removed. thanks for catching this